Non optional connector (#3129)
modelica/ModelicaSpecification: chapters/annotations.tex

\end{annotationdefinition}

> \section{Usage Restrictions}\label{usage-restrictions}

Reviewer: I would have have applied the table-based style of presentation to a new section like this, but  it might be better to include such as change in #2957…

Author: I think it is best to wait, especially as I separated the two annotations.

The annotation \fmtannotationindex{singleInstance} in a class indicates that there should only be one component instance of the class, and it should be in the same scope as the class is defined.
The intent is to remove the class when the component is removed and to prevent duplication of the component.

> \subsection{Connection Restrictions}\label{connection-restrictions}

Reviewer: Empty line after `\subsection` (as everywhere else):
```suggestion
\subsection{Connection Restrictions}\label{connection-restrictions}

```

A connector component declaration may have the following annotation:
\begin{lstlisting}[language=modelica]
annotation(mustBeConnected = "message");
\end{lstlisting}%
\annotationindex{mustBeConnected}

It makes it an error if the connector is not connected from the outside (for a conditional connector this check is only active if the connector is enabled).
For an array of connectors it applies separately to each element.

\begin{nonnormative}
This annotation is intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.
It is particularly suited for stream connectors, see \cref{stream-connectors}.
\end{nonnormative}

\begin{example}
This can be used for some optional connectors that should be connected when conditionally enabled.
\begin{lstlisting}[language=modelica]
partial model PartialWithSupport
  Flange_a flange_a;

>   Flange_b flange_b;

Reviewer: Maybe the listing should contain some equations that motivate having these connectors as part of the example?

Author: If we look at more complete models there are no such equations in the interface-classes.
Obviously they are used in the actual models - like a torque-sources.

However, having just an optional "support"-connector would look weird.

Author: I stripped it down to only one flange.

>   Support support($\ldots$)

Reviewer: Maybe there is no need to indicate elided code here?
```suggestion
  Support support
```

    if useSupport
    annotation(
      mustBeConnected = "Support connector should be connected if activated.");

>   $\ldots$

Reviewer: Maybe there's no need to indicate elided code here?
```suggestion
```

end PartialWithSupport;
\end{lstlisting}
\end{example}

A connector component declaration may have the following annotation:
\begin{lstlisting}[language=modelica]
annotation(mayOnlyConnectOnce = "message");
\end{lstlisting}%
\annotationindex{mayOnlyConnectOnce}

It makes it an error if the connector is connected from the outside and:
\begin{itemize}
\item For non-stream connectors the connection set has more than two elements.

> \item For stream connectors (see \cref{stream-connectors}), the connection set has more than two elements whose flow variable may be negative.

Reviewer: The explanation of this rule was given in https://github.com/modelica/ModelicaSpecification/pull/3129#discussion_r872366054, but I still think that a reader of the specification might also want to see the of explanation.

Author: This has now been added

> \item For stream connectors (see \cref{stream-connectors}), the connection set has more than two elements whose flow variable may be negative (based on the \lstinline!min!-attribute).

Reviewer: Is the error expected to trigger during translation, or is it expected that changing the `min` parameter at initialization time could resolve the situation?  If not (and short of `min` being a constant attribute), one could consider something more aggressive, like this:
```suggestion
\item For stream connectors (see \cref{stream-connectors}), the connection set has more than two elements whose flow variable may be negative (based on evaluation of the \lstinline!min!-attribute).
```

\end{itemize}
For an array of connectors it applies separately to each element.

\begin{nonnormative}
This annotation is intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.
The connection handling operates on connection sets, and thus this restriction should also operate on those sets.
The set handling avoids the case where only one of two equivalent models generate diagnostics.
The stream connector part is primarily intended to exclude sensor-variables, see \cref{connection-of-3-stream-connectors-where-one-mass-flow-rate-is-identical-to-zero-n-3-and}, but also excludes non-reversible outgoing flows.
\end{nonnormative}

\begin{example}
This can be used for components that implement mixing of fluids where it is not desired to combine that with the normal stream-connector mixing.
\begin{lstlisting}[language=modelica]
model MultiPort
  parameter Integer n = 0 annotation(Dialog(connectorSizing = true));
  FluidPort_a port_a(redeclare package Medium = Medium);
  FluidPorts_b ports_b[n](redeclare each package Medium = Medium)
    annotation (mayOnlyConnectOnce = "Should only connect once per element!");

>   $\ldots$

Reviewer: Similar to above, does it really add value to the example to indicate elided code here?

Reviewer: ```suggestion
```

Author: Ok, kaking it partial also suggest that something is missing without that.

\end{lstlisting}
\end{example}

A connector component declaration may have the following annotation:
\begin{lstlisting}[language=modelica]
annotation(mustBeConnected = "message");
\end{lstlisting}%
\annotationindex{mustBeConnected}

> If non-empty it indicates that it is an error if the connector is not connected from the outside (for a conditional connector this check is only active if the connector is enabled).

Reviewer: What would an empty annotation look like, and what would it mean?  Would it be better to just say the following?
```suggestion
It makes it an error if the connector is not connected from the outside (for a conditional connector this check is only active if the connector is enabled).
```

For an array of connectors it applies separately to each element.

\begin{nonnormative}

> Note that this annotation is only intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.

Reviewer: Redundant _Note_ inside non-normative section:
```suggestion
This annotation is only intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.
```

The _only intended_ almost sounds normative.  Could _only_ be omitted – it not, there seems to be something normative missing?
```suggestion
This annotation is intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.
```

> It is in particular suited for stream-connectors, \cref{stream-connectors}.

Reviewer: Language simplification:
```suggestion
It is particularly suited for stream-connectors, \cref{stream-connectors}.
```

\end{nonnormative}

\begin{example}
This can be used for some optional connectors that should be connected when conditionally enabled.
\begin{lstlisting}[language=modelica]
partial model PartialWithSupport
  Flange_a flange_a;
  Flange_b flange_b;
  Support support($\ldots$)

>     if useSupport annotation(mustBeConnected = "Support connector should be connected if activated.");

Reviewer: Line is too long:
```suggestion
    if useSupport
    annotation(
      mustBeConnected = "Support connector should be connected if activated.");
```

  $\ldots$
end PartialWithSupport;
\end{lstlisting}
\end{example}

A connector component declaration may have the following annotation:
\begin{lstlisting}[language=modelica]
annotation(mayOnlyConnectOnce = "message");
\end{lstlisting}%
\annotationindex{mayOnlyConnectOnce}

> If non-empty it indicates that it is an error if it is connected from the outside and:

Reviewer: Similar simplification:
```suggestion
It makes it an error if the connector is connected from the outside and:
```

\begin{itemize}
\item For non-stream connectors the connection set has more than two elements.

> \item For stream connectors, \cref{stream-connectors}, the connection set has more than two elements whose flow variable may be negative.

Reviewer: Insert _see_:
```suggestion
\item For stream connectors (see \cref{stream-connectors}), the connection set has more than two elements whose flow variable may be negative.
```

> \item For stream connectors, \cref{stream-connectors}, the connection set has more than two elements whose flow variable may be negative.

Reviewer: What does this restriction really mean? Is it a condition to be checked during simulation?

Author: During translation using min-values as in https://specification.modelica.org/master/derivation-of-stream-equations.html#connection-of-3-stream-connectors-where-one-mass-flow-rate-is-identical-to-zero-n-3-and

Reviewer: Outdated thread replaced by new comment. Resolving.

\end{itemize}
For an array of connectors it applies separately to each element.

\begin{nonnormative}

> The connection handling operates on connection sets, and thus this restriction should also operate on those sets.
> The set handling avoids the case where only one of two equivalent models generate diagnostics.
> The stream connector part is primarily intended to exclude sensor-variables, but also excludes non-reversible outgoing flows.
> Note that this annotation is only intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.

Reviewer: Similar simplification as above:
```suggestion
This annotation is intended for non-causal connectors, see \cref{restrictions-of-connections-and-connectors}.
The connection handling operates on connection sets, and thus this restriction should also operate on those sets.
The set handling avoids the case where only one of two equivalent models generate diagnostics.
The stream connector part is primarily intended to exclude sensor-variables, but also excludes non-reversible outgoing flows.
```

\end{nonnormative}

\begin{example}
This can be used for components that implement mixing of fluids where it is not desired to combine that with the normal stream-connector mixing.
\begin{lstlisting}[language=modelica]
model MultiPort

>   parameter Integer n=0 annotation(Dialog(connectorSizing=true));

Reviewer: Space:
```suggestion
  parameter Integer n = 0 annotation(Dialog(connectorSizing = true));
```

>   FluidPort_a port_a(redeclare package Medium=Medium);
>   FluidPorts_b ports_b[n](redeclare each package Medium=Medium)

Reviewer: Space:
```suggestion
  FluidPort_a port_a(redeclare package Medium = Medium);
  FluidPorts_b ports_b[n](redeclare each package Medium = Medium)
```

\end{lstlisting}
\end{example}

> 
> 

Reviewer: I guess one empty line should be sufficient here?
```suggestion

```

Author: Ok.
